utils/android/adb/commands.py

```diff
@@ -0,0 +1,151 @@
+
+    1. 'adb shell' can only take input up to a certain length, so it fails for
+       long executable names or when a large amount of arguments are passed to
+       the executable. This function attempts to limit the size of any string
+       passed to 'adb shell'.
+    2. 'adb shell' ignores the exit code of any command it runs. This function
+       therefore uses its own mechanisms to determine whether the executable
+       had a successful exit code when run on device.
+
+    If 'rerun_failures' is specified, this function re-runs failing tests,
+    printing their output to stdout.
+    """
+    # We'll be running the executable in a temporary directory in
+    # /data/local/tmp. `adb shell` has trouble with commands that
+    # exceed a certain length, so to err on the safe side we only
+    # use the first 10 characters of the UUID.
+    uuid_dir = '{}/{}'.format(DEVICE_TEMP_DIR, str(uuid.uuid4())[:10])
+    shell(['mkdir', '-p', uuid_dir])
+
+    # `adb` can only handle commands under a certain length. No matter what the
+    # original executable's name, on device we call it `__executable`.
+    executable = '{}/__executable'.format(uuid_dir)
+    push(executable_path, executable)
+
+    # When running the executable on the device, we need to pass it the same
+    # arguments, as well as specify the correct LD_LIBRARY_PATH. Save these
+    # to a file we can easily call multiple times.
+    executable_with_args = '{}/__executable_with_args'.format(uuid_dir)
+    _create_executable_on_device(
+        executable_with_args,
+        'LD_LIBRARY_PATH={uuid_dir}:{tmp_dir} '
+        '{executable} {executable_arguments}'.format(
+            uuid_dir=uuid_dir,
+            tmp_dir=DEVICE_TEMP_DIR,
+            executable=executable,
+            executable_arguments=' '.join(executable_arguments)))
+
+    # Write the output from the test executable to a file named '__stdout', and
+    # if the test executable succeeds, write 'SUCCEEDED' to a file
+    # named '__succeeded'. We do this because `adb shell` does not report
+    # the exit code of the command it executes on the device, so instead we
+    # check the '__succeeded' file for our string.
+    executable_stdout = '{}/__stdout'.format(uuid_dir)
+    succeeded_token = 'SUCCEEDED'
+    executable_succeeded = '{}/__succeeded'.format(uuid_dir)
+    executable_piped = '{}/__executable_piped'.format(uuid_dir)
+    _create_executable_on_device(
+        executable_piped,
+        '{executable_with_args} > {executable_stdout} && '
+        'echo "{succeeded_token}" > {executable_succeeded}'.format(
+            executable_with_args=executable_with_args,
+            executable_stdout=executable_stdout,
+            succeeded_token=succeeded_token,
+            executable_succeeded=executable_succeeded))
+
+    # We've pushed everything we need to the device.
+    # Now execute the wrapper script.
+    shell([executable_piped])
```

Comment:
This script does a few things:

1. For each test, create a temporary directory on the device, like `/data/local/tmp/ef6763e7`, and pushes the test executables to that directory.
2. Runs the test.
3. **If the test succeeded**, delete the entire directory -- in other words, `rm -rf /data/local/tmp/ef6763e7`. I originally decided to do this in order to debug test failures: by preserving the products on the phone, I could run commands like `adb cat /data/local/tmp/ef6763e7/out` to see what had happened.

On top of these steps, `command_upload_stdlib` pushes build products to the root test directory (`/data/local/tmp` by default).

I agree that we want to use a clean device for each test run. A few questions, though:

1. `command_upload_stdlib` only uploads, it doesn't delete existing stdlib build products on the device. So if I upload `foo.framework` and `bar.framework`, then modify the source code such that `bar.framework` is no longer built, then upload the stdlib again, the device will still have both `foo.framework` and `bar.framework`. Should I add some logic to `command_upload_stdlib` to clean the device each time?
2. Since each test is run in a separate directory with a UUID, they shouldn't conflict (and thus should be "clean" runs). I originally decided not to delete temporary directories if the tests failed, in order to make debugging easier. Still, it is true that, over many test runs, this could cause a device to fill its hard drive with test artifacts. Should I modify this to always remove temporary directories?